### PR TITLE
[retiarii] fix experiment does not exit after done

### PR DESCRIPTION
**Pull request by QuanluZhang**

### CODE REVIEW

File: test/ut/retiarii/test_multitrial.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+exp_config = RetiariiExeConfig('local')
+exp_config.experiment_name = 'mnist_unittest'
+exp_config.trial_concurrency = 2
+exp_config.max_trial_number = 2
+exp_config.training_service.use_active_gpu = False
+exp.run(exp_config, 8080)
+assert isinstance(exp.export_top_models()[0], dict)
+"""
+
+def test_exp_exit_without_stop():
```
Comment (ultmaster) [inline test/ut/retiarii/test_multitrial.py:53]: You can write a function creating this experiment and use `multiprocessing.Process(target=my_testing_experiment).start()` to launch the sub-process.

File: test/ut/retiarii/test_multitrial.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    script_name = 'tmp_multi_trial.py'
+    with open(script_name, 'w') as f:
+        f.write(python_script)
+    proc = Popen(['python3', script_name])
```
Comment (ultmaster) [inline test/ut/retiarii/test_multitrial.py:57]: It's not always python3.

File: nni/runtime/msg_dispatcher_base.py [hunk 3]
```diff
@@ -46,8 +46,8 @@ def __init__(self, command_channel_url=None):
             self._channel.connect()
         self.default_command_queue = Queue()
         self.assessor_command_queue = Queue()
-        self.default_worker = threading.Thread(target=self.command_queue_worker, args=(self.default_command_queue,))
-        self.assessor_worker = threading.Thread(target=self.command_queue_worker, args=(self.assessor_command_queue,))
+        self.default_worker = threading.Thread(target=self.command_queue_worker, args=(self.default_command_queue,), daemon=True)
```
Comment (ultmaster) [inline nni/runtime/msg_dispatcher_base.py:49]: Please add comments to explain why daemon=True, and what problem it fixes.

File: test/ut/retiarii/test_oneshot.py [hunk 4]
```diff
@@ -216,13 +217,13 @@ def _mnist_net(type_, evaluator_kwargs):
...
-    train_dataset = MNIST('data/mnist', train=True, download=True, transform=transform)
+    train_dataset = nni.trace(MNIST)('data/mnist', train=True, download=True, transform=transform)
     # Multi-GPU combined dataloader will break this subset sampler. Expected though.
-    train_random_sampler = RandomSampler(train_dataset, True, int(len(train_dataset) / 20))
-    train_loader = DataLoader(train_dataset, 64, sampler=train_random_sampler)
-    valid_dataset = MNIST('data/mnist', train=False, download=True, transform=transform)
-    valid_random_sampler = RandomSampler(valid_dataset, True, int(len(valid_dataset) / 20))
-    valid_loader = DataLoader(valid_dataset, 64, sampler=valid_random_sampler)
+    train_random_sampler = nni.trace(RandomSampler)(train_dataset, True, int(len(train_dataset) / 20))
+    train_loader = nni.trace(DataLoader)(train_dataset, 64, sampler=train_random_sampler)
```
Comment (ultmaster) [inline test/ut/retiarii/test_oneshot.py:223]: DataLoader is already wrapped. No need to wrap again.

File: test/ut/retiarii/test_multitrial.py (hunk 2 above)
Comment (QuanluZhang) [reply test/ut/retiarii/test_multitrial.py:57]: yeah, any suggestion for this?

**Review by ultmaster — APPROVED** [review]

File: test/ut/retiarii/test_oneshot.py (hunk 4 above)
Comment (QuanluZhang) [reply test/ut/retiarii/test_oneshot.py:223]: yeah, it is pretty hard to know which one should be wrapped... so, I wrapped them all for simplicity.

File: nni/runtime/msg_dispatcher_base.py (hunk 3 above)
Comment (QuanluZhang) [reply nni/runtime/msg_dispatcher_base.py:49]: added

File: test/ut/retiarii/test_multitrial.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+exp.run(exp_config, 8080)
+assert isinstance(exp.export_top_models()[0], dict)
+"""
+
+def test_exp_exit_without_stop():
+    script_name = 'tmp_multi_trial.py'
+    with open(script_name, 'w') as f:
+        f.write(python_script)
+    proc = Popen([sys.executable, script_name])
+    proc.wait()
```
Comment (J-shang) [inline test/ut/retiarii/test_multitrial.py:62]: should we add a max execute duration here?

File: test/ut/retiarii/test_multitrial.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        search_strategy = strategy.Random()
+        exp = RetiariiExperiment(base_model, evaluator, strategy=search_strategy)
+        exp_config = RetiariiExeConfig('local')
+        exp_config.experiment_name = 'mnist_unittest'
+        exp_config.trial_concurrency = 1
+        exp_config.max_trial_number = 1
+        exp_config.training_service.use_active_gpu = False
+        exp.run(exp_config, 8080)
+        assert isinstance(exp.export_top_models()[0], dict)
+        exp.stop()
```
Comment (J-shang) [inline test/ut/retiarii/test_multitrial.py:38]: experiment can quickly release the port after `stop()`? (I don't know, maybe `local` don't have to wait for the port release)

**Review by J-shang — APPROVED** [review]

File: test/ut/retiarii/test_multitrial.py (hunk 6 above)
Comment (QuanluZhang) [reply test/ut/retiarii/test_multitrial.py:38]: according to my test, it can.

File: test/ut/retiarii/test_multitrial.py (hunk 5 above)
Comment (QuanluZhang) [reply test/ut/retiarii/test_multitrial.py:62]: what is setting max execution duration mainly used for?

**Pull request merged**